tests/Core/Ruleset/PropertyTypeHandlingTest.php

```diff
@@ -38,6 +38,35 @@ final class PropertyTypeHandlingTest extends TestCase
     const SNIFF_CLASS = 'Fixtures\\TestStandard\\Sniffs\\SetProperty\\PropertyTypeHandlingSniff';
 
 
+    /**
+     * Verify a deprecation notice is shown when an array property is set from the ruleset using a comma-separated string.
+     *
+     * Support for this format was (soft) deprecated in PHPCS 3.3.0.
+     *
+     * @return void
+     */
+    public function testUsingOldSchoolArrayFormatShowsDeprecationNotice()
+    {
+        $regex  = '`^(';
+        $regex .= 'DEPRECATED: Passing an array of values to a property using a comma-separated string\R';
+        $regex .= 'was deprecated in PHP_CodeSniffer 3\.3\.0\. Support will be removed in PHPCS 4\.0\.0\.\R';
+        $regex .= 'The deprecated syntax was used for property "expectsOldSchool(?:EmptyArray|ArrayWith(?:Extended|Only)?(?:KeysAnd)?Values)"\R';
+        $regex .= 'for sniff "';
+        $regex .= '(?:\./tests/Core/Ruleset/Fixtures/TestStandard/Sniffs/SetProperty/PropertyTypeHandlingSniff\.php|TestStandard\.SetProperty\.PropertyTypeHandling)';
+        $regex .= '"\.\R';
+        $regex .= 'Pass array values via <element \[key="\.\.\." \]value="\.\.\."> nodes instead\.\R';
+        $regex .= '){14}\R$`';
```

Comment:
Is it intentional that the deprecation notice for each property is displayed twice? There are seven properties using the old array format in `PropertyTypeHandlingTest.xml`. So, I expected this regex group to repeat seven times, not fourteen.

On a quick check, it seems this happens only when a sniff path is passed to `<rule ref=''>`. It doesn't happen when a sniff code is passed.

When a sniff path is passed, `$todo` ends up with two entries for the same sniff (one with the path and another with the code), and then the code that adds the deprecation notice for a given property is executed twice:

https://github.com/PHPCSStandards/PHP_CodeSniffer/blob/f2274bd522558da42af39572adfccc48e8737f83/src/Ruleset.php#L1078-L1100

Reply:
@rodrigoprimo You're 100% correct and I agree, it's not pretty to have those messages twice. I did consider filtering on sniff code, but ended up deciding against it for the following reasons:
1. If people include by file name, they may not be aware of the sniff code, so filtering out the file name messages would make it harder for them to identify what to fix in their ruleset.
2. If the sniff referenced by file doesn't follow the PHPCS naming conventions, the sniff code may not be descriptive at all.

Reply:
That makes sense, and I'm okay with leaving the duplication for the reasons you mentioned.

Reply:
We could maybe add a code comment documenting that the deprecated messages will be displayed twice when a sniff is included by file name.

Or check if it is easy to detect that the sniff was included by file name and then include the sniff path to the deprecation message and display it only once.

That being said, since those deprecation messages will be removed when 4.0 is released, I don't think those suggestions need to be addressed.

Reply:
> That being said, since those deprecation messages will be removed when 4.0 is released, I don't think those suggestions need to be addressed.

Exactly. This code should only be in the codebase for a short time anyway.

Reply:
Having said that, the same issue will happen when the deprecation becomes an error in PHPCS 4.0, but hopefully people will never see that as they've fixed up their ruleset ahead of time (based on the deprecation).

Still, the same arguments against filtering out the duplicate messages would apply for 4.0.

Reply:
Good point about the problem still existing in PHPCS 4.0 as the deprecations will become errors. I haven't considered it. I'm okay with merging this PR as is.

I just wondered if you considered the option below. Maybe it is not easy to implement so not worth it?

> Or check if it is easy to detect that the sniff was included by file name and then include the sniff path to the deprecation message and display it only once.

Reply:
I don't think this would be easy to implement (otherwise I'd have done it already). It would require some kind of keeping track of which deprecations have been registered already and for something which has been deprecated for 8 years now and will soon no longer be supported, I don't think it's worth the effort.

Reply:
:shipit: